src/analysis/processing/qgsalgorithmvirtualrastercalculator.cpp

```diff
@@ -0,0 +1,177 @@
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmvirtualrastercalculator.h"
+#include "qgsrasterdataprovider.h"
+
+///@cond PRIVATE
+
+QgsProcessingAlgorithm::Flags QgsVirtualRasterCalculatorAlgorithm::flags() const
+{
+  return QgsProcessingAlgorithm::flags() | QgsProcessingAlgorithm::FlagNoThreading;
+}
+
+QString QgsVirtualRasterCalculatorAlgorithm::name() const
+{
+  return QStringLiteral( "virtualrastercalc" );
+}
+
+QString QgsVirtualRasterCalculatorAlgorithm::displayName() const
+{
+  return QObject::tr( "Raster calculator (virtual)" );
+}
+
+QStringList QgsVirtualRasterCalculatorAlgorithm::tags() const
+{
+  return QObject::tr( "raster,calculator,virtual" ).split( ',' );
+}
+
+QString QgsVirtualRasterCalculatorAlgorithm::group() const
+{
+  return QObject::tr( "Raster analysis" );
+}
+
+QString QgsVirtualRasterCalculatorAlgorithm::groupId() const
+{
+  return QStringLiteral( "rasteranalysis" );
+}
+
+QString QgsVirtualRasterCalculatorAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Performs algebraic operations using raster layers and generates in-memory result." );
+}
+
+QgsVirtualRasterCalculatorAlgorithm *QgsVirtualRasterCalculatorAlgorithm::createInstance() const
+{
+  return new QgsVirtualRasterCalculatorAlgorithm();
+}
+
+void QgsVirtualRasterCalculatorAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "INPUT" ), QObject::tr( "Input layers" ), QgsProcessing::SourceType::TypeRaster ) );
+  addParameter( new QgsProcessingParameterExpression( QStringLiteral( "EXPRESSION" ), QObject::tr( "Expression" ), QVariant(),  QStringLiteral( "INPUT" ), false, Qgis::ExpressionType::RasterCalculator ) );
+  addParameter( new QgsProcessingParameterExtent( QStringLiteral( "EXTENT" ), QObject::tr( "Output extent" ), QVariant(), true ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "CELL_SIZE" ), QObject::tr( "Output cell size (leave empty to set automatically)" ), QgsProcessingParameterNumber::Double, QVariant(), true, 0.0 ) );
+  addParameter( new QgsProcessingParameterCrs( QStringLiteral( "CRS" ), QObject::tr( "Output CRS" ), QVariant(), true ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "LAYER_NAME" ), QObject::tr( "Output layer name" ), QVariant(), false, true ) );
+  addOutput( new QgsProcessingOutputRasterLayer( QStringLiteral( "OUTPUT" ), QObject::tr( "Calculated" ) ) );
+}
+
+QVariantMap QgsVirtualRasterCalculatorAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+
+  const QList< QgsMapLayer * > layers = parameterAsLayerList( parameters, QStringLiteral( "INPUT" ), context );
```

Comment:
filter out non-raster layers here?

Reply:
Filtering already done at the multiple layers parameter side, it should be safe.